ConnectionRefused uten detaljer på feilkonfigurerte discovery-URL-er

Får stacktraces á la: 
```
Caused by: no.nav.security.oidc.exceptions.MetaDataNotAvailableException: java.net.ConnectException: Connection refused: connect
	at no.nav.security.oidc.configuration.IssuerConfiguration.getProviderMetadata(IssuerConfiguration.java:114)
	at no.nav.security.oidc.configuration.IssuerConfiguration.<init>(IssuerConfiguration.java:40)
...
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source)
	at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
...
```
når Spring Boot-properties har feil i issuer-discovery-URL-ene. F.eks. følgende konfigurasjon:
```
no.nav.security.oidc.issuer.localhost.discoveryurl=http://no-such-host/metadata
no.nav.security.oidc.issuer.localhost.acceptedaudience=aud-localhost
```

Her bør unntaket `ConnectException` fanges opp og dekoreres med detaljer om i det minste vertsnavnet, gjerne også hele URL-en og/eller annen aktuell informasjon om utstederen.

--- a/token-validation-core/src/main/java/no/nav/security/token/support/core/exceptions/MetaDataNotAvailableException.java
+++ b/token-validation-core/src/main/java/no/nav/security/token/support/core/exceptions/MetaDataNotAvailableException.java
@@ -17,7 +17,7 @@ public class MetaDataNotAvailableException extends RuntimeException {
     }
 
     public MetaDataNotAvailableException(URL url, Exception e) {
-        this("Kunne ikke hente metadata fra " + url, e);
+        this(String.format("could not retrieve metadata from url: %s. received exception %s", url, e), e);
     }
 
 }

--- a/token-validation-core/src/test/java/no/nav/security/token/support/core/IssuerMockWebServer.java
+++ b/token-validation-core/src/test/java/no/nav/security/token/support/core/IssuerMockWebServer.java
@@ -9,11 +9,11 @@ import okhttp3.mockwebserver.MockResponse;
 import okhttp3.mockwebserver.MockWebServer;
 import okhttp3.mockwebserver.RecordedRequest;
 import okio.BufferedSink;
+import org.jetbrains.annotations.NotNull;
 
 import java.io.IOException;
 import java.net.URI;
 import java.net.URL;
-import java.net.URLEncoder;
 import java.nio.charset.StandardCharsets;
 
 @Slf4j
@@ -24,19 +24,31 @@ public class IssuerMockWebServer {
     private MockWebServer proxyServer;
     private URL discoveryUrl;
     private URL proxyUrl;
+    private final boolean startProxyServer;
+
+    public IssuerMockWebServer(){
+        this(true);
+    }
+
+    public IssuerMockWebServer(boolean startProxyServer){
+        this.startProxyServer = startProxyServer;
+    }
 
     public void start() throws IOException {
         this.server = new MockWebServer();
         this.server.start();
         this.discoveryUrl = this.server.url(DISCOVERY_PATH).url();
         this.server.setDispatcher(new Dispatcher() {
+            @NotNull
             @Override
-            public MockResponse dispatch(RecordedRequest request) {
-                log.info("received request on url={} with headers={}", request.getRequestUrl(), request.getHeaders());
-                log.debug("path():{} compared to {}", request.getRequestUrl().encodedPath(), DISCOVERY_PATH);
+            public MockResponse dispatch(@NotNull RecordedRequest request) {
+                log.debug("received request on url={} with headers={}", request.getRequestUrl(), request.getHeaders());
+                log.debug("comparing path in request '{}' with '{}'", request.getRequestUrl().encodedPath(), DISCOVERY_PATH);
                 if (request.getRequestUrl().encodedPath().endsWith(DISCOVERY_PATH)) {
+                    log.debug("returning well-known json data");
                     return wellKnownJson();
                 } else {
+                    log.error("path not found, returning 404");
                     return new MockResponse().setResponseCode(404);
                 }
             }
@@ -44,8 +56,10 @@ public class IssuerMockWebServer {
 
         this.proxyServer = new MockWebServer();
         this.proxyServer.setDispatcher(new ProxyDispatcher(HttpUrl.parse(discoveryUrl.toString())));
-        this.proxyServer.start();
-        this.proxyUrl = URI.create("http://localhost:" + this.proxyServer.getPort()).toURL();
+        if(startProxyServer){
+            this.proxyServer.start();
+            this.proxyUrl = URI.create("http://localhost:" + this.proxyServer.getPort()).toURL();
+        }
     }
 
     public void shutdown() throws IOException {
@@ -79,6 +93,7 @@ public class IssuerMockWebServer {
             client = new OkHttpClient.Builder().build();
         }
 
+        @NotNull
         @Override
         public MockResponse dispatch(final RecordedRequest recordedRequest) {
             Request.Builder requestBuilder = new Request.Builder()
@@ -94,7 +109,7 @@ public class IssuerMockWebServer {
                     }
 
                     @Override
-                    public void writeTo(BufferedSink sink) throws IOException {
+                    public void writeTo(@NotNull BufferedSink sink) throws IOException {
                         recordedRequest.getBody().clone().readAll(sink);
                     }
 
@@ -105,24 +120,27 @@ public class IssuerMockWebServer {
                 });
             }
             Request request = requestBuilder.build();
-            log.info("created request to destination: {}", request);
+            log.debug("created request to destination: {}", request);
             try (Response response = client.newCall(request).execute()) {
                 ResponseBody body = response.body();
                 if (body != null) {
-                    return new MockResponse()
-                        .setHeaders(response.headers())
-                        .setBody(body.string())
-                        .setResponseCode(response.code());
+                    MockResponse mockResponse = new MockResponse();
+                    mockResponse.headers(response.headers());
+                    mockResponse.setBody(body.string());
+                    mockResponse.setResponseCode(response.code());
+                    return mockResponse;
                 } else {
-                    return new MockResponse()
-                        .setStatus("proxy error, response body from destination was null")
-                        .setResponseCode(500);
+                    MockResponse mockResponse = new MockResponse();
+                    mockResponse.status("proxy error, response body from destination was null");
+                    mockResponse.setResponseCode(500);
+                    return mockResponse;
                 }
             } catch (IOException e) {
                 log.error("got exception when proxying request.", e);
-                return new MockResponse()
-                    .setStatus("proxy error: " + e.getMessage())
-                    .setResponseCode(500);
+                MockResponse mockResponse = new MockResponse();
+                mockResponse.status("proxy error: " + e.getMessage());
+                mockResponse.setResponseCode(500);
+                return mockResponse;
             }
         }
     }

--- a/token-validation-core/src/test/java/no/nav/security/token/support/core/configuration/IssuerConfigurationTest.java
+++ b/token-validation-core/src/test/java/no/nav/security/token/support/core/configuration/IssuerConfigurationTest.java
@@ -1,0 +1,66 @@
+package no.nav.security.token.support.core.configuration;
+
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+import no.nav.security.token.support.core.IssuerMockWebServer;
+import no.nav.security.token.support.core.exceptions.MetaDataNotAvailableException;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+class IssuerConfigurationTest {
+
+    private IssuerMockWebServer issuerMockWebServer;
+
+    @BeforeEach
+    void setup() throws IOException {
+        issuerMockWebServer = new IssuerMockWebServer(false);
+        issuerMockWebServer.start();
+    }
+
+    @AfterEach
+    void after() throws IOException {
+        issuerMockWebServer.shutdown();
+    }
+
+    @Test
+    void issuerConfigurationWithMetadataFromDiscoveryUrl() {
+        IssuerConfiguration config = new IssuerConfiguration(
+            "issuer1", issuerMockWebServer.getDiscoveryUrl(), List.of("audience1"), new ProxyAwareResourceRetriever());
+        assertThat(config.getMetaData()).isNotNull();
+        assertThat(config.getTokenValidator()).isNotNull();
+        OIDCProviderMetadata metadata = config.getMetaData();
+        assertThat(metadata.getIssuer()).isNotNull();
+        assertThat(metadata.getJWKSetURI()).isNotNull();
+    }
+
+    @Test
+    void issuerConfigurationDiscoveryUrlNotValid() throws MalformedURLException {
+        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> {
+            new IssuerConfiguration(
+                "issuer1",
+                URI.create("http://notvalid").toURL(), List.of("audience1"),
+                new ProxyAwareResourceRetriever());
+        });
+        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> {
+            new IssuerConfiguration(
+                "issuer1",
+                URI.create("http://localhost").toURL(), List.of("audience1"),
+                new ProxyAwareResourceRetriever());
+        });
+        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> {
+            new IssuerConfiguration(
+                "issuer1",
+                URI.create(issuerMockWebServer.getDiscoveryUrl().toString() + "/pathincorrect").toURL(),
+                List.of("audience1"),
+                new ProxyAwareResourceRetriever());
+        });
+    }
+}

--- a/token-validation-core/src/test/java/no/nav/security/token/support/core/configuration/MultiIssuerConfigurationTest.java
+++ b/token-validation-core/src/test/java/no/nav/security/token/support/core/configuration/MultiIssuerConfigurationTest.java
@@ -1,5 +1,6 @@
 package no.nav.security.token.support.core.configuration;
 
+import com.nimbusds.jose.util.DefaultResourceRetriever;
 import lombok.extern.slf4j.Slf4j;
 import no.nav.security.token.support.core.IssuerMockWebServer;
 import org.junit.jupiter.api.AfterEach;
@@ -34,7 +35,6 @@ class MultiIssuerConfigurationTest {
 
     @Test
     void getIssuerConfiguration() {
-
         IssuerProperties issuerProperties = new IssuerProperties(discoveryUrl, List.of("audience1"));
         String issuerName = "issuer1";
         MultiIssuerConfiguration multiIssuerConfiguration =
@@ -44,7 +44,6 @@ class MultiIssuerConfigurationTest {
 
     @Test
     void getIssuerConfigurationWithProxy() {
-
         IssuerProperties issuerProperties = new IssuerProperties(discoveryUrl, List.of("audience1"));
         issuerProperties.setProxyUrl(proxyUrl);
         String issuerName = "issuer1";
@@ -53,9 +52,9 @@ class MultiIssuerConfigurationTest {
 
         assertThatMultiIssuerConfigurationIsPopulatedFromMetadata(issuerName, multiIssuerConfiguration);
         IssuerConfiguration config = multiIssuerConfiguration.getIssuer(issuerName).orElse(null);
-
-
-        System.out.println("proxy: " + ((ProxyAwareResourceRetriever) config.getResourceRetriever()).getProxy());
+        assertThat(config).isNotNull();
+        assertThat(config.getResourceRetriever()).isInstanceOf(DefaultResourceRetriever.class);
+        assertThat(((DefaultResourceRetriever) config.getResourceRetriever()).getProxy()).isNotNull();
     }
 
     private void assertThatMultiIssuerConfigurationIsPopulatedFromMetadata(String issuerName,

